crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,527 @@
+                    }
+                }
+
+                // main loop
+                let world = SyncUnsafeCell::from_mut(world);
+                while self.completed_systems.count_ones(..) != self.completed_systems.len() {
+                    if !self.exclusive_running {
+                        self.spawn_system_tasks(scope, schedule, world);
+                    }
+
+                    if self.running_systems.count_ones(..) > 0 {
+                        // wait for systems to complete
+                        let index = self
+                            .receiver
+                            .recv()
+                            .await
+                            .unwrap_or_else(|error| unreachable!("{}", error));
+
+                        self.finish_system_and_signal_dependents(index);
+
+                        while let Ok(index) = self.receiver.try_recv() {
+                            self.finish_system_and_signal_dependents(index);
+                        }
+
+                        self.rebuild_active_access();
+                    }
+                }
+
+                // SAFETY: all systems have completed
+                let world = unsafe { &mut *world.get() };
+                Self::apply_system_buffers(&mut self.unapplied_systems, schedule, world);
+
+                debug_assert_eq!(self.ready_systems.count_ones(..), 0);
+                debug_assert_eq!(self.running_systems.count_ones(..), 0);
+                debug_assert_eq!(self.unapplied_systems.count_ones(..), 0);
+                self.active_access.clear();
+                self.completed_sets.clear();
+                self.completed_systems.clear();
+            };
+
+            #[cfg(feature = "trace")]
+            let runner_span = info_span!("schedule_task");
+            #[cfg(feature = "trace")]
+            let runner = runner.instrument(runner_span);
+            scope.spawn(runner);
+        });
+    }
+}
+
+impl MultiThreadedExecutor {
+    pub fn new() -> Self {
+        let (sender, receiver) = async_channel::unbounded();
+        Self {
+            system_task_metadata: Vec::new(),
+            sender,
+            receiver,
+            dependents_scratch: Vec::new(),
+            active_access: default(),
+            local_thread_running: false,
+            exclusive_running: false,
+            completed_sets: FixedBitSet::new(),
+            completed_systems: FixedBitSet::new(),
+            ready_systems: FixedBitSet::new(),
+            seen_ready_systems: FixedBitSet::new(),
+            running_systems: FixedBitSet::new(),
+            unapplied_systems: FixedBitSet::new(),
+        }
+    }
+
+    fn spawn_system_tasks<'scope>(
+        &mut self,
+        scope: &Scope<'_, 'scope, ()>,
+        schedule: &'scope SystemSchedule,
+        world: &'scope SyncUnsafeCell<World>,
+    ) {
+        while let Some(system_index) = self
+            .ready_systems
+            .difference(&self.seen_ready_systems)
+            .next()
```

Comment:
So the `while` loop is there because `ready_systems` is mutated inside `should_run` when systems are skipped. 

Since that restarts the iteration, `seen_ready_systems` is there to avoid calling `can_run` multiple times for the same system (its result won't change).

Reply:
The bitset `difference` is "accelerated" (just ANDs the underlying `usize` blocks) but we can do the one iteration over `ready_systems` and just `continue` if the index is in `completed` or in `seen_ready`.

Reply:
Made a change along these lines.